File: src/zoho_books_cli/commands/recurring_expenses.py
"""`zb recurring-expenses ...` — full coverage of /recurring-expenses.

Thin wrappers: CRUD + stop/resume + children/history. Stop and resume use PUT
per Zoho's API spec.
"""

from __future__ import annotations

import typer

from zoho_books_cli import config
from zoho_books_cli.client import ZohoBooksClient
from zoho_books_cli.commands import _shared

app = typer.Typer(
    help="Recurring expense operations (CRUD + stop/resume + children/history).",
    no_args_is_help=True,
)

BASE = "/recurringexpenses"


@app.command("list")
def list_recurring(
    query: list[str] = typer.Option(
        None, "--query", "-q", help="Query params as key=value. May be repeated."
    ),
    params: str = typer.Option(
        None,
        "--params",
        help="Query params as a JSON object. Merged on top of --query.",
    ),
    page: int = typer.Option(None, "--page", help="Page number (1-indexed)."),
    per_page: int = typer.Option(None, "--per-page", help="Rows per page."),
    page_all: bool = typer.Option(
        False, "--page-all", help="Auto-paginate (NDJSON: one page per line)."
    ),
    page_limit: int = typer.Option(10, "--page-limit", help="Max pages with --page-all."),
    page_delay: int = typer.Option(
        100, "--page-delay", help="Delay between pages in ms with --page-all."
    ),
):
    """List recurring expenses. Returns one page plus page_context."""
    q = _shared.parse_query_pairs(query, params)
    if page is not None:
        q["page"] = str(page)
    if per_page is not None:
        q["per_page"] = str(per_page)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        _shared.emit_list_paginated(
            client,
            BASE,
            q,
            "recurring_expenses",
            page_all=page_all,
            page_limit=page_limit,
            page_delay_ms=page_delay,
        )


@app.command("create")
def create_recurring(
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Create a recurring expense."""
    json_body = _shared.parse_body(body)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(BASE, json_body=json_body)
    _shared.emit_object(resp)


@app.command("get")
def get_recurring(
    recurring_expense_id: str = typer.Argument(..., help="Zoho Books recurring_expense_id."),
):
    """Get a single recurring expense by ID."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.get(f"{BASE}/{recurring_expense_id}")
    _shared.emit_object(resp)


@app.command("update")
def update_recurring(
    recurring_expense_id: str = typer.Argument(..., help="Zoho Books recurring_expense_id."),
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Update a recurring expense by ID."""
    json_body = _shared.parse_body(body)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.put(f"{BASE}/{recurring_expense_id}", json_body=json_body)
    _shared.emit_object(resp)


@app.command("update-by-custom-field")
def update_recurring_by_custom_field(
    key: str = typer.Option(..., "--key", help="Custom-field API name."),
    value: str = typer.Option(..., "--value", help="Custom-field value to match on."),
    body: str = typer.Option(
        ..., "--body", "-b", help="JSON body with the update fields. IDs must be strings."
    ),
    upsert: bool = typer.Option(
        False, "--upsert", help="Create a new recurring expense if no match is found."
    ),
):
    """Update a recurring expense by a custom field's unique value.

    Identifier is sent via X-Unique-Identifier-Key / X-Unique-Identifier-Value headers
    per Zoho's spec; the request body carries only the fields to update.
    """
    json_body = _shared.parse_body(body)
    headers = {"X-Unique-Identifier-Key": key, "X-Unique-Identifier-Value": value}
    if upsert:
        headers["X-Upsert"] = "true"
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.put(BASE, json_body=json_body, headers=headers)
    _shared.emit_object(resp)


@app.command("delete")
def delete_recurring(
    recurring_expense_id: str = typer.Argument(..., help="Zoho Books recurring_expense_id."),
):
    """Delete a recurring expense by ID."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.delete(f"{BASE}/{recurring_expense_id}")
    _shared.emit_action("recurring_expense_id", recurring_expense_id, resp)


@app.command("stop")
def stop_recurring(
    recurring_expense_id: str = typer.Argument(..., help="Zoho Books recurring_expense_id."),
):
    """Stop a recurring expense (POST /recurringexpenses/{id}/status/stop)."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(f"{BASE}/{recurring_expense_id}/status/stop")
    _shared.emit_action("recurring_expense_id", recurring_expense_id, resp)


@app.command("resume")
def resume_recurring(
    recurring_expense_id: str = typer.Argument(..., help="Zoho Books recurring_expense_id."),
):
    """Resume a recurring expense (POST /recurringexpenses/{id}/status/resume)."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(f"{BASE}/{recurring_expense_id}/status/resume")
    _shared.emit_action("recurring_expense_id", recurring_expense_id, resp)


@app.command("children")
def list_children(
    recurring_expense_id: str = typer.Argument(..., help="Zoho Books recurring_expense_id."),
    query: list[str] = typer.Option(None, "--query", "-q", help="Query params as key=value."),
    params: str = typer.Option(
        None,
        "--params",
        help="Query params as a JSON object. Merged on top of --query.",
    ),
    page: int = typer.Option(None, "--page", help="Page number (1-indexed)."),
    per_page: int = typer.Option(None, "--per-page", help="Rows per page."),
    page_all: bool = typer.Option(
        False, "--page-all", help="Auto-paginate (NDJSON: one page per line)."
    ),
    page_limit: int = typer.Option(10, "--page-limit", help="Max pages with --page-all."),
    page_delay: int = typer.Option(
        100, "--page-delay", help="Delay between pages in ms with --page-all."
    ),
):
    """List child expenses created from a recurring expense."""
    q = _shared.parse_query_pairs(query, params)
    if page is not None:
        q["page"] = str(page)
    if per_page is not None:
        q["per_page"] = str(per_page)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        _shared.emit_list_paginated(
            client,
            f"{BASE}/{recurring_expense_id}/expenses",
            q,
            "expenses",
            page_all=page_all,
            page_limit=page_limit,
            page_delay_ms=page_delay,
        )


@app.command("history")
def list_history(
    recurring_expense_id: str = typer.Argument(..., help="Zoho Books recurring_expense_id."),
):
    """List history / comments for a recurring expense."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.get(f"{BASE}/{recurring_expense_id}/comments")
    _shared.emit_list(resp, "comments")

File: src/commands/recurring_expenses.rs
//! `zb recurring-expenses` — CRUD + stop/resume + children + history over
//! `/recurringexpenses`. Stop/resume use POST /status/{stop|resume}.

use clap::{Args, Subcommand};

use crate::cli::Ctx;
use crate::commands::common::{self, BodyArgs, CustomFieldUpdateArgs, ListArgs};
use crate::errors::Result;

const BASE: &str = "/recurringexpenses";
const ID_FIELD: &str = "recurring_expense_id";

#[derive(Args, Debug)]
pub struct Cmd {
    #[command(subcommand)]
    pub sub: Sub,
}

#[derive(Subcommand, Debug)]
pub enum Sub {
    /// List recurring expenses.
    List(ListArgs),
    /// Create a recurring expense.
    Create(BodyArgs),
    /// Get a single recurring expense by ID.
    Get(IdArgs),
    /// Update a recurring expense by ID.
    Update(UpdateArgs),
    /// Update a recurring expense by a custom field's unique value.
    #[command(name = "update-by-custom-field")]
    UpdateByCustomField(CustomFieldUpdateArgs),
    /// Delete a recurring expense by ID.
    Delete(IdArgs),
    /// Stop a recurring expense.
    Stop(IdArgs),
    /// Resume a recurring expense.
    Resume(IdArgs),
    /// List child expenses created from a recurring expense.
    Children(ChildrenArgs),
    /// List history / comments for a recurring expense.
    History(IdArgs),
}

#[derive(Args, Debug)]
pub struct IdArgs {
    /// Zoho Books recurring_expense_id.
    pub recurring_expense_id: String,
}

#[derive(Args, Debug)]
pub struct UpdateArgs {
    pub recurring_expense_id: String,
    #[command(flatten)]
    pub body: BodyArgs,
}

#[derive(Args, Debug)]
pub struct ChildrenArgs {
    pub recurring_expense_id: String,
    #[command(flatten)]
    pub list: ListArgs,
}

pub fn run(cmd: Cmd, ctx: &mut Ctx) -> Result<()> {
    match cmd.sub {
        Sub::List(args) => common::list(ctx, BASE, &args, "recurring_expenses"),
        Sub::Create(args) => common::create(ctx, BASE, &args),
        Sub::Get(args) => common::get(ctx, &format!("{BASE}/{}", args.recurring_expense_id)),
        Sub::Update(args) => common::update(
            ctx,
            &format!("{BASE}/{}", args.recurring_expense_id),
            &args.body,
        ),
        Sub::UpdateByCustomField(args) => common::update_custom(ctx, BASE, &args),
        Sub::Delete(args) => {
            let path = format!("{BASE}/{}", args.recurring_expense_id);
            common::delete(ctx, &path, ID_FIELD, &args.recurring_expense_id)
        }
        Sub::Stop(args) => {
            let path = format!("{BASE}/{}/status/stop", args.recurring_expense_id);
            common::action(ctx, &path, ID_FIELD, &args.recurring_expense_id)
        }
        Sub::Resume(args) => {
            let path = format!("{BASE}/{}/status/resume", args.recurring_expense_id);
            common::action(ctx, &path, ID_FIELD, &args.recurring_expense_id)
        }
        Sub::Children(args) => {
            let path = format!("{BASE}/{}/expenses", args.recurring_expense_id);
            common::list(ctx, &path, &args.list, "expenses")
        }
        Sub::History(args) => {
            let path = format!("{BASE}/{}/comments", args.recurring_expense_id);
            let resp = ctx.client.get(&path, &crate::shared::Query::new())?;
            common::emit_list_flat(&resp, "comments", ctx)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::cli::Ctx;

    #[test]
    fn list_targets_recurringexpenses() {
        let mut server = mockito::Server::new();
        let m = server
            .mock("GET", "/books/v3/recurringexpenses")
            .match_query(mockito::Matcher::Any)
            .with_status(200)
            .with_body(r#"{"recurring_expenses":[]}"#)
            .create();
        let mut ctx = Ctx::new_for_test(&server.url());
        run(
            Cmd {
                sub: Sub::List(ListArgs::default()),
            },
            &mut ctx,
        )
        .unwrap();
        m.assert();
    }
}
